File: python/caesar.py
import argparse
import sys

upper = list('ABCDEFGHIJKLMNOPQRSTUVWXYZÆØÅ')
lower = list('abcdefghijklmnopqrstuvwxyzæøå')
size  = len(upper)

# ~i encodes lowercase: ~0=-1, ~1=-2, ... mirrors the Go/JS implementations
idx: dict[str, int] = {ch: i for i, ch in enumerate(upper)}
idx.update({ch: ~i for i, ch in enumerate(lower)})


def cipher(text: str, shift: int) -> str:
    shift = (shift % size + size) % size
    out = []
    for ch in text:
        pos = idx.get(ch)
        if pos is None:
            out.append(ch)
        elif pos >= 0:
            out.append(upper[(pos + shift) % size])
        else:
            out.append(lower[(~pos + shift) % size])
    return ''.join(out)


def main() -> None:
    parser = argparse.ArgumentParser(prog='caesar')
    parser.add_argument('shift', type=int)
    parser.add_argument('file')
    parser.add_argument('-d', '--decrypt', action='store_true')
    parser.add_argument('-o', '--output')
    args = parser.parse_args()

    try:
        text = open(args.file, encoding='utf-8').read()
    except OSError as e:
        print(f'could not read file: {e}', file=sys.stderr)
        sys.exit(1)

    result = cipher(text, -args.shift if args.decrypt else args.shift)

    if args.output:
        try:
            with open(args.output, 'w', encoding='utf-8') as f:
                f.write(result)
        except OSError as e:
            print(f'could not write file: {e}', file=sys.stderr)
            sys.exit(1)
    else:
        sys.stdout.write(result)


if __name__ == '__main__':
    main()

File: bun/caesar.ts
const upper = [...'ABCDEFGHIJKLMNOPQRSTUVWXYZÆØÅ'];
const lower = [...'abcdefghijklmnopqrstuvwxyzæøå'];
const size  = upper.length;

const idx = new Map<string, number>([
  ...upper.map((ch, i): [string, number] => [ch, i]),
  ...lower.map((ch, i): [string, number] => [ch, ~i]),
]);

export function cipher(text: string, shift: number): string {
  shift = ((shift % size) + size) % size;
  return [...text].map(ch => {
    const pos = idx.get(ch);
    if (pos === undefined) return ch;
    return pos >= 0
      ? upper[(pos + shift) % size]
      : lower[(~pos + shift) % size];
  }).join('');
}

if (import.meta.main) {
  const args = Bun.argv.slice(2);
  if (args.length < 2) {
    process.stderr.write('usage: caesar <shift> <file> [-d|--decrypt] [-o|--output <file>]\n');
    process.exit(1);
  }

  const shift = parseInt(args[0], 10);
  if (isNaN(shift)) {
    process.stderr.write(`invalid shift: ${args[0]}\n`);
    process.exit(1);
  }

  const inputFile = args[1];
  let outputFile: string | null = null;
  let decrypt = false;

  for (let i = 2; i < args.length; i++) {
    switch (args[i]) {
      case '-d':
      case '--decrypt':
        decrypt = true;
        break;
      case '-o':
      case '--output':
        if (i + 1 >= args.length) {
          process.stderr.write('missing value for -o/--output\n');
          process.exit(1);
        }
        outputFile = args[++i];
        break;
      default:
        process.stderr.write(`unknown flag: ${args[i]}\n`);
        process.exit(1);
    }
  }

  const text = await Bun.file(inputFile).text();
  const result = cipher(text, decrypt ? -shift : shift);

  if (outputFile) {
    await Bun.write(outputFile, result);
  } else {
    process.stdout.write(result);
  }
}
